### ISSUE TITLE
RFC: Customize accessor names

### ISSUE THREAD

**Issue by schani** [quicktype/quicktype#500]
# Motivation

Users of quicktype have requested finer control over how properties and accessors are named in the generated code.  In the case of class properties and enum members this can be because the names in an existing schema are bad but cannot be changed, or because they want to avoid quicktype renaming the properties, which it might do in cases where property names conflict with keywords, for example.  For union members there is currently no way to specify names at all, but it is a useful feature to have.

The [JSON Schema Core specification](http://json-schema.org/latest/json-schema-core.html) explicitly allows applications to define their own extensions, both for the purposes of validation as well as to add annotations that the application can use for its purposes.

# Proposal

We propose adding a property `accessors` to `object` types as well as to types that have an `enum` or a `oneOf` property, i.e. enum and union types.  The property has the following type:

* For object types it is an object whose keys are the object property names and whose values are accessor entries.

* For types with an `enum` property it is an object whose keys are the enumeration members and whose values are accessor entries.

* For types with a `oneOf` property it is an array consisting of accessor entries, each corresponding to the type in the `oneOf` array at the same index.

Accessor entries have one of these types.

* A string, giving the user's name for that property/enum member/union case for all target languages.

* An object whose keys are target language names and whose values are string, giving the user's name for that property/enum member/union case for that specific programming language.  The key `*` matches all target languages not otherwise named.

# Examples

Simple cases for customizing properties, enum members, and union
cases, for all programming languages at once:

```json
{
    "type": "object",
    "properties": {
        "union": {
            "oneOf": [
                { "type": "boolean" },
                { "type": "number" }
            ],
            "accessors": [
                "presence",
                "amount"
            ]
        },
        "enum": {
            "type": "string",
            "enum": ["red", "green", "blue"],
            "accessors": {
                "red": "fire",
                "green": "grass"
            }
        }
    },
    "accessors": {
        "union": "unionization",
        "enum": "enumerification"
    }
}
```

The properties `union` and `enum` of the top-level class will be called `unionization` and `enumerification`, respectively, in all target languages.  The `boolean` union member will be called `presence` and the `number` member will be called `amount`.  The `enum` members `red` and `green` will be called `fire` and `grass`, respectively, while the `blue` member will not be modified.

This shows how to customize per programming language:

```json
{
    "type": "object",
    "properties": {
        "foo": { "type": "string" },
        "bar": { "type": "string" }
    },
    "accessors": {
        "foo": {
            "csharp": "fu",
            "java": "goo"
        },
        "bar": {
            "go": "bah",
            "*": "barre"
        }
    }
}
```

Here the property `foo` will be called `fu` in C#, `goo` in Java, and for all other languages it stays `foo`.  The property `bar` will be called `bah` in Go, and `barre` in all other target languages.

# Extensibility

If we need to add more information in addition to the accessor names, such as whether to make the property immutable, we can allow accessor entries to be objects, not just strings, but only if `accessors` is an object with target languages (otherwise it would be ambigious).  So, for example, you might be able to specify to rename the `foo` property and make it immutable like this (this format is not part of the proposal!):

```json
    "accessors": {
        "foo": {
            "*": {
                "name": "fu",
                "immutable": true
            }
        }
    }
```

**Comment by schani** [comment]
My first draft did use an object for the enums, then I changed it to an array because the enumeration specifies an order, too, but you're right, it's probably better to make it an object.  It has the added benefit that it's robust against changing the order of the enumeration.  I'll change it in the proposal.

Drawbacks?  My main concerns are designing it in a way that properly addresses the problems and that's future proof.  I don't see any drawbacks in the ability to arbitrarily choose accessor names, other than the problem that any kind of freedom can be abused by inexperienced developers.  And note that your case (exact control and backward compatibility) is only one of the two I'm trying to address: some people have to deal with really shitty JSON that comes from a service that takes data out of a 50 year old mainframe database where the keys are 6 characters long and all caps.  They need some help, too ;-)

**Comment by schani** [comment]
I've implemented a prototype of this in #684 .  I changed the name of the `accessors` property to `qt-accessors`.  [This is an example](https://github.com/quicktype/quicktype/blob/accessor-names/test/inputs/schema/accessors.schema).

@josefalcon Please let me know what you think.  The main thing I'm unsure about is what to do with name styles.  Right now the assigned names are put through name styling just like all other names.  My feeling tells me that we should do name styling for assigned names except when you assign a name for a target language explicitly.  I.e. if you assign the property name `my_property` without specifying target languages, or via the wildcard, you would get `MyProperty` in C#.  Only if you assign it specifically to C# would it ignore the styling and make it `my_property` (this is not implemented yet).  Thoughts?
